src/Jaeger.Core/SpanContext.cs

```diff
@@ -1,54 +1,132 @@
+            SpanId parentId,
+            SpanContextFlags flags,
+            ReadOnlyDictionary<string, string> baggage,
+            string debugId)
+        {
+            TraceId = traceId;
             SpanId = spanId;
+            ParentId = parentId;
             Flags = flags;
-            _baggage = baggage ?? new Dictionary<string, string>();
+            Baggage = baggage ?? throw new ArgumentNullException(nameof(baggage));
+            DebugId = debugId;
+        }
+
+        public IEnumerable<KeyValuePair<string, string>> GetBaggageItems()
+        {
+            return Baggage;
+        }
+
+        public string GetBaggageItem(string key)
+        {
+            return Baggage.TryGetValue(key, out string value) ? value : null;
         }
 
-        public bool IsSampled => Flags.HasFlag(ContextFlags.Sampled);
+        public string ContextAsString()
+        {
+            return $"{TraceId}:{SpanId}:{ParentId}:{((byte)Flags).ToString("x")}";
+        }
 
-        public override string ToString() {
-            return $"{TraceId}:{SpanId}:{ParentId}:{(byte)Flags}";
+        public override string ToString()
+        {
+            return ContextAsString();
         }
 
-        public static SpanContext FromString(string from) {
-            if (from == string.Empty) { throw new Exception("Cannot convert empty string to SpanContext"); }
+        public static SpanContext ContextFromString(string value)
+        {
+            if (string.IsNullOrEmpty(value))
+                throw new ArgumentException($"Cannot convert empty string to {nameof(SpanContext)}");
 
-            var parts = from.Split(':');
-            if (parts.Length != 4) { throw new Exception("String does not match tracer state format"); }
+            var parts = value.Split(':');
+            if (parts.Length != 4)
+                throw new ArgumentException("String does not match tracer state format");
 
-            return new SpanContext(TraceId.FromString(parts[0]), SpanId.FromString(parts[1]), SpanId.FromString(parts[2]), null, (ContextFlags)byte.Parse(parts[3]));
+            return new SpanContext(
+                TraceId.FromString(parts[0]),
+                SpanId.FromString(parts[1]),
+                SpanId.FromString(parts[2]),
+                (SpanContextFlags)byte.Parse(parts[3], NumberStyles.HexNumber));
         }
 
-        // OpenTracing API: Iterate through all baggage items
-        public IEnumerable<KeyValuePair<string, string>> GetBaggageItems()
+        public SpanContext WithBaggageItem(string key, string value)
+        {
+            var newBaggage = new Dictionary<string, string>(Baggage);
+            if (value == null)
+            {
+                newBaggage.Remove(key);
+            }
+            else
+            {
+                newBaggage[key] = value;
+            }
+            return new SpanContext(TraceId, SpanId, ParentId, Flags, new ReadOnlyDictionary<string, string>(newBaggage), DebugId);
```

Comment:
Use `IReadOnlyDictionary`. `Dictionary` already implements `IReadOnlyDictionary`. Wrapping in `ReadOnlyDictionary` not needed.